upload/views/test_results.py

```diff
@@ -0,0 +1,116 @@
+import logging
+import uuid
+
+from rest_framework import serializers, status
+from rest_framework.exceptions import NotAuthenticated
+from rest_framework.permissions import BasePermission
+from rest_framework.response import Response
+from rest_framework.views import APIView
+
+from codecov_auth.authentication.repo_auth import (
+    OrgLevelTokenAuthentication,
+    RepositoryLegacyTokenAuthentication,
+)
+from codecov_auth.authentication.types import RepositoryAsUser
+from codecov_auth.models import Owner, Service
+from core.models import Commit
+from reports.models import CommitReport
+from services.archive import ArchiveService
+from services.redis_configuration import get_redis_connection
+from upload.helpers import dispatch_upload_task
+from upload.serializers import FlagListField
+from upload.views.helpers import get_repository_from_string
+
+log = logging.getLogger(__name__)
+
+
+class UploadTestResultsPermission(BasePermission):
+    def has_permission(self, request, view):
+        return request.auth is not None and "upload" in request.auth.get_scopes()
+
+
+class UploadSerializer(serializers.Serializer):
+    commit = serializers.CharField(required=True)
+    slug = serializers.CharField(required=True)
+    build = serializers.CharField(required=False)
+    buildURL = serializers.CharField(required=False)
+    job = serializers.CharField(required=False)
+    flags = FlagListField(required=False)
+    pr = serializers.CharField(required=False)
+    service = serializers.CharField(required=False)
+
+
+class TestResultsView(APIView):
+    permission_classes = [UploadTestResultsPermission]
+    authentication_classes = [
+        OrgLevelTokenAuthentication,
+        RepositoryLegacyTokenAuthentication,
+    ]
+
+    def post(self, request):
+        serializer = UploadSerializer(data=request.data)
+        if not serializer.is_valid():
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+        data = serializer.validated_data
+
+        if isinstance(request.user, Owner):
+            # using org token
+            owner = request.user
+            repo = get_repository_from_string(Service(owner.service), data["slug"])
+        elif isinstance(request.user, RepositoryAsUser):
+            # repository token
+            repo = request.user._repository
+        else:
+            raise NotAuthenticated()
+
+        commit, _ = Commit.objects.get_or_create(
+            commitid=data["commit"],
+            repository=repo,
+            defaults={
+                "branch": data.get("branch"),
+                "pullid": data.get("pr"),
+                "merged": False if data.get("pr") is not None else None,
+                "state": "pending",
+            },
+        )
+
+        upload_external_id = str(uuid.uuid4())
+
+        storage_path = f"v1/uploads/{upload_external_id}.json"
```

Comment:
[nit] I think there should be some indication in the path that this is a TestResult report. For bundle_analysis the indication is in the bucket (it's a [special bucket](https://github.com/codecov/shared/blob/main/shared/bundle_analysis/storage.py#L12)). But it seems that this one is the same as the coverage reports.

My suggestion is the obvious `test_results/v1/uploads/{upload_external_id}.json`

Personally I'd keep the coverage report's style of adding a repo hash and commit info as well, cause it makes it a tad easier to know that I'm looking at the right report if I have to debug something, but that is just my personal preference and maybe the extra effort is not worth it